### PR TITLE
Fix precision of the alternative routes fork intersection visualization

### PR DESCRIPTION
**Pull request by LukasPaczos**
### Description

This PR fixes an issue, where the alternative routes vanishing point would be shifted, depending on the route's length, from the actual location of the fork. The problem has been fixed by adopting the [EPSG:3857 projection](https://epsg.io/3857) for calculating the distance and distance remaining of the alternative routes to derive an offset at which they should vanish (that is exactly on the fork with the primary route).

This PR also refactors the generation and caching of points used for this distance calculation, that is also shared with the primary line's continuous vanishing of the traveled portion of the route. I noticed a very significant improvement in the generation by swapping the previously used `SparseArray` for a regular primitive array. For long routes, there were so many resize operations needed to be done internally by the `SparseArray` implementation that it added up. The performance wins on this front allowed to simplify all the threading previously needed to initialize the vanishing route line component. 

And one last addition is a tests that verify, that Nav Native is appending a random ID to routes that do not have the server-side request UUID. This is necessary for caching purposes, since this PR introduce a new, more performant cache-key that is based only on the route ID, for use-cases that do not require the exact equality.

### Screenshots or Gifs

| Before  | After |
| ------------- | ------------- |
| Screenshot_20220825_133303 | Screenshot_20220825_133123 |
|  | Screenshot_20220825_133136 |

| Before  | After |
| ------------- | ------------- |
| Screenshot_20220825_133408 | Screenshot_20220825_133523 |
|  | Screenshot_20220825_133552 |

### CODE REVIEW

File: libnavui-maps/api/current.txt [hunk 1]
```diff
@@ -1184,17 +1184,6 @@ package com.mapbox.navigation.ui.maps.route.line.model {
     method public com.mapbox.maps.extension.style.expressions.generated.Expression generateExpression();
   }
 
-  public final class RouteLineGranularDistances {
```
Comment (LukasPaczos) [inline libnavui-maps/api/current.txt:1187]: I'm intentionally removing this data class from the public API. It should've ever been exposed, and it has no practical usage, it's only consumed by internal Nav SDK APIs.

File: libnavui-maps/api/current.txt [hunk 2]
```diff
@@ -1294,17 +1283,6 @@ package com.mapbox.navigation.ui.maps.route.line.model {
     property public final Throwable? throwable;
   }
 
-  public final class RoutePoints {
```
Comment (LukasPaczos) [inline libnavui-maps/api/current.txt:1297]: Same as above.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/util/CacheResultUtils.kt [hunk 3]
```diff
@@ -17,6 +19,38 @@ internal object CacheResultUtils {
         override fun invoke(f: (P1, P2) -> R) = f(p1, p2)
     }
 
+    /**
+     * Specialized cache key that can be used for managing results of processing related to a route.
+     *
+     * This key uses [NavigationRoute.id] instead of [NavigationRoute.hashCode] because the latter produces a hash of the full [DirectionsRoute]
+     * which for longer routes and use cases that require very frequent access (multiple times per second) proves to be inefficient (refs https://github.com/mapbox/navigation-sdks/issues/1918).
```
Comment (dzinad) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/util/CacheResultUtils.kt:26]: It's an internal issue, is it OK to place the link here?

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtils.kt [hunk 4]
```diff
@@ -63,13 +63,28 @@ internal object MapboxRouteLineUtils {
...
 
     private val extractRouteDataCache: LruCache<
         CacheResultUtils.CacheResultKey2<
             DirectionsRoute, (RouteLeg) -> List<String>?,
             List<ExtractedRouteData>
             >,
-        List<ExtractedRouteData>> by lazy { LruCache(3) }
+        List<ExtractedRouteData>> by lazy { LruCache(NUMBER_OF_SUPPORTED_ROUTES) }
+
+    internal val routePointsCache: LruCache<
```
Comment (cafesilencio) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtils.kt:75]: Does this need to be internal or can it be made private?

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtils.kt [hunk 5]
```diff
@@ -63,13 +63,28 @@ internal object MapboxRouteLineUtils {
...
+        CacheResultUtils.CacheResultKeyRoute<RoutePoints?>, RoutePoints?>
+        by lazy { LruCache(NUMBER_OF_SUPPORTED_ROUTES) }
+
+    internal val granularDistancesCache: LruCache<
```
Comment (cafesilencio) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtils.kt:79]: Can/should this be private?

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtils.kt [hunk 6]
```diff
@@ -63,13 +63,28 @@ internal object MapboxRouteLineUtils {
...
+        CacheResultUtils.CacheResultKeyRoute<
+            RouteLineGranularDistances?>, RouteLineGranularDistances?>
+        by lazy { LruCache(NUMBER_OF_SUPPORTED_ROUTES) }
+
+    internal val distinctGranularDistancesCache: LruCache<
```
Comment (cafesilencio) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtils.kt:84]: Can/should this be private?

**Review by cafesilencio — APPROVED** [review]
LGTM

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/util/CacheResultUtils.kt (hunk 3 above)
Comment (LukasPaczos) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/util/CacheResultUtils.kt:26]: That's okay, it's an internal class and the repo itself has a generic name.

**Pull request merged**